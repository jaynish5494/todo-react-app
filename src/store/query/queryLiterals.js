export const queryLiterals = name => ({
  REQUEST: `${name}_REQUEST`,
  SUCCESS: `${name}_SUCCESS`,
  ERROR: `${name}_ERROR`
});

export const actionCreator = type => values => ({
  type,
  ...values
});

export const queryActions = name => {
  if (typeof name !== 'string' && !(name instanceof String)) {
    throw new Error('query action creators expect string for initialisation');
  }

  const { REQUEST, SUCCESS, ERROR } = queryLiterals(name);

  return {
    request: actionCreator(REQUEST),
    success: actionCreator(SUCCESS),
    error: actionCreator(ERROR)
  };
};
