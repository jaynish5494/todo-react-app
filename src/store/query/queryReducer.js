import { queryLiterals } from './queryLiterals';

export const initialState = {
  loading: false,
  error: null,
  data: []
};

export const createQueryReducer = name => {
  return function queryReducer(state = initialState, action) {
    const { REQUEST, SUCCESS, ERROR } = queryLiterals(name);
    const { data, error, type } = action;
    switch (type) {
      case REQUEST:
        return {
          ...state,
          loading: true,
          error: null
        };
      case SUCCESS:
        return {
          ...state,
          loading: false,
          data
        };
      case ERROR:
        return {
          ...state,
          loading: false,
          error
        };
      default:
        return state;
    }
  };
};
