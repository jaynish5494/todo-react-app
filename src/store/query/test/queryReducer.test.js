import { initialState, createQueryReducer } from '../queryReducer';
import { queryLiterals } from '../queryLiterals';

const name = 'TEST';

describe('query reducer creator', () => {
  it('should have a well defined initial state for query reducer creation.', () => {
    expect(initialState).toMatchSnapshot();
  });

  it('should create query reducers which use the default initial state if no state is defined.', () => {
    const expected = initialState;
    const actual = createQueryReducer(name);

    expect(actual(undefined, {})).toEqual(expected);
  });
});

describe('should handle all actions', () => {
  const mockQueryLiterals = jest.fn();
  mockQueryLiterals.mockImplementation(() => queryLiterals(name));
  const { REQUEST, SUCCESS, ERROR } = mockQueryLiterals();
  it('should only handle REQUEST, SUCCESS, ERROR actions', () => {
    const sut = createQueryReducer(name);

    const expected = initialState;
    const actual = sut(initialState, {
      type: 'ANY_OTHER_ACTION_TYPE'
    });

    expect(actual).toEqual(expected);
  });

  it('should have a case for REQUEST action & handle this case correctly.', () => {
    const sut = createQueryReducer(name);

    const expected = {
      ...initialState,
      loading: true,
      error: null
    };
    const actual = sut(initialState, {
      type: REQUEST
    });

    expect(actual).toEqual(expected);
  });

  it('should have a case for SUCCESS action & handle this case correctly.', () => {
    const sut = createQueryReducer(name);

    const data = ['value1', 'value2'];

    const expected = {
      ...initialState,
      data,
      error: null,
      loading: false
    };
    const actual = sut(initialState, {
      type: SUCCESS,
      data
    });

    expect(actual).toEqual(expected);
  });

  it('should have a case for ERROR action & handle this case correctly.', () => {
    const sut = createQueryReducer(name);

    const error = {
      msg: 'error'
    };

    const expected = {
      ...initialState,
      error,
      loading: false
    };
    const actual = sut(initialState, {
      type: ERROR,
      error
    });

    expect(actual).toEqual(expected);
  });
});
