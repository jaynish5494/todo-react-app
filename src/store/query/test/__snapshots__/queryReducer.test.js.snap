// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query reducer creator should have a well defined initial state for query reducer creation. 1`] = `
Object {
  "data": Array [],
  "error": null,
  "loading": false,
}
`;
